File: apps/faithform-android/app/src/main/kotlin/io/faithform/app/ui/groups/GroupInvitation.kt
package io.faithform.app.ui.groups

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.faithform.app.contract.*
import io.faithform.app.network.ApiClient
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch
import kotlinx.serialization.json.*

@Composable fun GroupInvitationScreen(api: ApiClient, token: String, onDismiss: () -> Unit, onJoined: (String) -> Unit) {
    val store = remember(token) { GroupsStore(api, "") }; val scope = rememberCoroutineScope()
    var preview by remember(token) { mutableStateOf<GroupInvitationPreview?>(null) }; var accepted by remember(token) { mutableStateOf(false) }; var retry by remember { mutableIntStateOf(0) }
    LaunchedEffect(token, retry) { try { store.error = null; preview = store.send("api/mobile/v1/group-invitations/preview", buildJsonObject { put("token", token) }) } catch (e: CancellationException) { throw e } catch (e: Exception) { store.error = GroupsStore.message(e) } }
    GroupFormSheet("You’re invited", onDismiss) {
        preview?.let { invitation ->
            Box(Modifier.fillMaxWidth().height(190.dp), contentAlignment = Alignment.Center) {
                GroupAvatarBackdrop(invitation.coverImageUrl, Modifier.matchParentSize())
                GroupAvatar(invitation.coverImageUrl, invitation.groupName, 104.dp)
            }
            Text(if (accepted) "Welcome to the group." else "There’s a place for you.", style = MaterialTheme.typography.headlineLarge)
            Text(invitation.groupName, style = MaterialTheme.typography.titleLarge)
            Text("You’re invited to join ${invitation.churchName}’s group. You’ll need to belong to this church in the app before joining.")
            Button(enabled = !store.busy, onClick = {
                if (accepted) onJoined(invitation.churchSlug)
                else scope.launch { store.action("Welcome to the group!") { val result = store.send<GroupJoinResult>("api/mobile/v1/group-invitations/accept", buildJsonObject { put("token", token) }); if (result.outcome !in listOf("joined", "already_member")) throw IllegalStateException(GroupsStore.outcomeMessage(result.outcome)); accepted = true } }
            }, modifier = Modifier.fillMaxWidth()) { Text(if (accepted) "Go to my groups" else if (store.busy) "Joining…" else "Accept invitation") }
        } ?: if (store.error == null) CircularProgressIndicator() else TextButton(onClick = { retry++ }) { Text("Try again") }
        GroupFeedback(store)
    }
}

File: apps/faithform-ios/App/Groups/GroupInvitation.swift
import SwiftUI
import FaithFormKit

struct GroupInvitationView: View {
    let api: APIClient
    let token: String
    let joined: (String) -> Void
    @Environment(\.dismiss) private var dismiss
    @State private var preview: GroupInvitationPreview?
    @State private var error: String?
    @State private var busy = false
    @State private var retry = 0
    @State private var accepted = false
    var body: some View {
        NavigationStack {
            ScrollView { VStack(alignment: .leading, spacing: 24) {
                if let preview {
                    ZStack {
                        GroupAvatarBackdrop(url: preview.coverImageUrl)
                        GroupAvatarView(url: preview.coverImageUrl, name: preview.groupName, size: 112)
                    }
                    .frame(height: 200).frame(maxWidth: .infinity)
                    .clipShape(RoundedRectangle(cornerRadius: 24, style: .continuous))
                    Text(accepted ? "Welcome to the group." : "There’s a place for you.").font(.largeTitle.bold())
                    Text(preview.groupName).font(.title2.bold())
                    Text("You’re invited to join \(preview.churchName)’s group. You’ll need to belong to this church in the app before joining.").foregroundStyle(.secondary)
                    if accepted { Button("Go to my groups") { joined(preview.churchSlug) }.buttonStyle(.borderedProminent) }
                    else { Button(busy ? "Joining…" : "Accept invitation") { Task { await accept() } }.buttonStyle(.borderedProminent).disabled(busy) }
                } else if error == nil { GroupDetailSkeleton() }
                if let error { Text(error).foregroundStyle(.red); if preview == nil { Button("Try again") { retry += 1 } } }
            }.padding(24) }.navigationTitle("You’re invited").navigationBarTitleDisplayMode(.inline)
                .toolbar { ToolbarItem(placement: .cancellationAction) { Button("Close") { dismiss() } } }
                .task(id: retry) { do { error = nil; let result = try await api.send("api/mobile/v1/group-invitations/preview", method: .post, body: GroupInvitationTokenRequest(token: token), as: GroupInvitationPreview.self); preview = result.value } catch is CancellationError {} catch { self.error = GroupsModel.message(error) } }
        }
    }
    private func accept() async {
        busy = true; error = nil; defer { busy = false }
        do {
            let result = try await api.send("api/mobile/v1/group-invitations/accept", method: .post, body: GroupInvitationTokenRequest(token: token), as: GroupJoinResult.self)
            guard let value = result.value, ["joined", "already_member"].contains(value.outcome) else { error = GroupsModel.outcomeMessage(result.value?.outcome ?? "unavailable"); return }
            accepted = true
        } catch is CancellationError {} catch { self.error = GroupsModel.message(error) }
    }
}
